filecoin/store/piece.js

```diff
@@ -0,0 +1,315 @@
+    group: encodedRecord.group,
+    insertedAt: encodedRecord.insertedAt,
+    updatedAt: encodedRecord.updatedAt
+  }
+}
+
+/**
+ * @param {PieceStoreRecordStatus} status 
+ * @returns {PieceStatus}
+ */
+const decodeStatus = (status) => {
+  switch (status) {
+    case Status.SUBMITTED: {
+      return 'submitted'
+    }
+    case Status.ACCEPTED: {
+      return 'accepted'
+    }
+    case Status.INVALID: {
+      return 'invalid'
+    }
+    default: {
+      throw new Error('invalid status received for decoding')
+    }
+  }
+}
+
+/**
+ * @param {PieceRecordQuery} recordKey 
+ */
+const encodeQueryProps = (recordKey) => {
+  return {
+    IndexName: 'stat',
+    KeyConditions: {
+      stat: {
+        ComparisonOperator: 'EQ',
+        AttributeValueList: [{ N: `${encodeStatus(recordKey.status)}` }]
+      }
+    }
+  }
+}
+
+
+/**
+ * Abstraction layer to handle operations on Piece Table.
+ *
+ * @param {string} region
+ * @param {string} tableName
+ * @param {object} [options]
+ * @param {string} [options.endpoint]
+ * @returns {import('@web3-storage/filecoin-api/storefront/api').PieceStore}
+ */
+export function createPieceTable (region, tableName, options = {}) {
+  const dynamoDb = new DynamoDBClient({
+    region,
+    endpoint: options.endpoint
+  })
+
+  return usePieceTable(dynamoDb, tableName)
+}
+
+/**
+ * @param {DynamoDBClient} dynamoDb
+ * @param {string} tableName
+ * @returns {import('@web3-storage/filecoin-api/storefront/api').PieceStore}
+ */
+export function usePieceTable(dynamoDb, tableName) {
+  return {
+    put: async (record) => {
+      const cmd = new PutItemCommand({
+        TableName: tableName,
+        Item: marshall(encodeRecord(record), {
+          removeUndefinedValues: true
+        }),
+
+      })
+
+      try {
+        await dynamoDb.send(cmd)
+      } catch (/** @type {any} */ error) {
+        return {
+          error: new StoreOperationFailed(`failed put to dynamo piece table, content: ${record.content}, piece: ${record.piece}`, { cause: error })
+        }
+      }
+
+      return {
+        ok: {}
+      }
+    },
+    get: async (key) => {
+      const getCmd = new GetItemCommand({
+        TableName: tableName,
+        Key: marshall(encodeKey(key)),
+      })
+      let res
+      try {
+        res = await dynamoDb.send(getCmd)
+      } catch (/** @type {any} */ error) {
+        if (error?.$metadata?.httpStatusCode === 404) {
+          return {
+            error: new RecordNotFound('item not found in store')
+          }
+        }
+        return {
+          error: new StoreOperationFailed(error.message)
+        }
+      }
+
+      // not found error
+      if (!res.Item) {
+        return {
+          error: new RecordNotFound('item not found in store')
+        }
+      }
+
+      return {
+        ok: decodeRecord(
+          /** @type {PieceStoreRecord} */ (unmarshall(res.Item))
+        )
+      }
+    },
+    has: async (key) => {
+      const getCmd = new GetItemCommand({
```

Comment:
```suggestion
      const getCmd = new HeadItemCommand({
```

Reply:
this is [dynamo](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-dynamodb/), there is no Head command

Reply:
haha sorry!